### PR TITLE
feat(core): add find_first_of methods to FixString class

### PR DESCRIPTION
**Pull request by dkrutskikh**
### **PR Type**
Enhancement

___

### **Description**
- Add `find_first_of` method with multiple overloads

- Support FixString, StringLike, C-string, and character searches

- Include comprehensive test coverage with edge cases

- Add private helper method for common search logic

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["FixString Class"] --> B["find_first_of overloads"]
  B --> C["FixString parameter"]
  B --> D["StringLike parameter"]
  B --> E["C-string parameter"]
  B --> F["char parameter"]
  G["_find_first_of_raw helper"] --> B
  H["Comprehensive tests"] --> B
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>fix_string.hpp</strong><dd><code>Add find_first_of method declarations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/fix_string.hpp

<ul><li>Add four <code>find_first_of</code> method overloads with documentation<br> <li> Support FixString, StringLike, C-string, and character parameters<br> <li> Add private <code>_find_first_of_raw</code> helper method declaration</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/88/files#diff-2ef8f9f195db23c05d195ee56d0aaa0c714eeaf911132af27804a48fa895ce44">+95/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>fix_string.inl</strong><dd><code>Implement find_first_of method logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/fix_string.inl

<ul><li>Implement four <code>find_first_of</code> method overloads<br> <li> Add <code>_find_first_of_raw</code> helper with optimized search logic<br> <li> Use <code>strpbrk</code> for multi-character searches, direct comparison for single <br>characters</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/88/files#diff-a47dfdd2a0becc17f40499abd90d00211a258ffaea3ce144fd5fc11d5ad08afd">+48/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>fix_string_test.cpp</strong><dd><code>Add find_first_of tests and cleanup</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/core/fix_string_test.cpp

<ul><li>Add comprehensive test suite for <code>find_first_of</code> functionality<br> <li> Test all parameter types, edge cases, and position handling<br> <li> Remove redundant comments in existing tests</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/88/files#diff-cc1ff39c1921067e1cd9ffcf0bc8c8f3d7349f0178f1eb6ddeac099bc0a2d2ae">+173/-6</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

## Greptile Summary

Updated On: 2025-09-06 20:06:19 UTC

This PR implements the `find_first_of` method for the FixString class, adding a critical string search functionality that was missing from the API. The implementation follows the established patterns in the codebase by providing four overloads to handle different input types: FixString objects, StringLike template types, C-style strings, and single characters.

The core functionality is implemented in the `_find_first_of_raw` helper method, which intelligently optimizes single-character searches by reusing the existing `_find_raw` method, while multi-character searches utilize `std::strpbrk` for efficient character set matching. This approach maintains consistency with other search methods like `find` and `rfind` that already exist in the class.

The PR also includes comprehensive test coverage with 167 new lines of test code covering various scenarios including different input types, edge cases with empty strings and position parameters, character sets, and special characters. Additionally, the changes include minor cleanup by removing redundant inline comments from existing tests, reducing code verbosity while maintaining clarity.

This addition brings the FixString class closer to feature parity with `std::string`, making it more suitable as a drop-in replacement for standard string operations in the ToyGine2 engine.

<details>
<summary>Changed Files</summary>

| Filename | Score | Overview |
|----------|-------|----------|
| `test/core/fix_string_test.cpp` | 5/5 | Adds comprehensive test coverage for `find_first_of` with 21 test scenarios and removes redundant comments |
| `include/core/fix_string.inl` | 4/5 | Implements the core `find_first_of` logic with optimized single/multi-character search handling |
| `include/core/fix_string.hpp` | 3/5 | Declares four `find_first_of` method overloads following standard C++ string API patterns |

</details>

## Confidence score: 3/5

- This PR requires careful review due to potential implementation inconsistencies and missing documentation
- Score lowered due to incomplete visibility of the `_find_first_of_raw` implementation and lack of PR description
- Pay close attention to the header/implementation alignment in `include/core/fix_string.hpp` and `include/core/fix_string.inl`

### What change is being made?
Add `find_first_of` methods to the `FixString` class to enable finding the first occurrence of any character from specified strings and account for multiple string types including `FixString`, `StringLike` objects, C strings, and individual characters.

### Why are these changes being made?
These changes enhance the `FixString` class by introducing functionality to locate the initial position of characters from various string formats within a `FixString`, which is a necessary feature for numerous string processing tasks. The implementation was guided by the need for case-sensitive searches and efficient handling of different input types.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`

### CODE REVIEW

File: include/core/fix_string.inl [hunk 1]
```diff
@@ -608,6 +608,33 @@ constexpr inline std::size_t FixString<allocatedSize>::rfind(char character, std
...
+                                                                     std::size_t position) const noexcept {
+  assert_message(string != nullptr, "String pointer must not be null");
+
+  return _find_first_of_raw(position, string, std::strlen(string));
+}
+
+template <std::size_t allocatedSize>
+constexpr inline std::size_t FixString<allocatedSize>::find_first_of(char character,
+                                                                     std::size_t position) const noexcept {
+  return _find_raw(position, &character, 1);
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/core/fix_string.inl:635]: won't be fixed

File: include/core/fix_string.inl [hunk 2]
```diff
@@ -608,6 +608,33 @@ constexpr inline std::size_t FixString<allocatedSize>::rfind(char character, std
...
+}
+
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/core/fix_string.inl:637]: won't be fixed
